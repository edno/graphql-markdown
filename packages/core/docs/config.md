# config

## Enumerations

### DeprecatedOption

#### Enumeration Members

##### DEFAULT

```ts
DEFAULT: "default";
```

###### Source

[config.ts:33](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L33)

##### GROUP

```ts
GROUP: "group";
```

###### Source

[config.ts:34](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L34)

##### SKIP

```ts
SKIP: "skip";
```

###### Source

[config.ts:35](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L35)

***

### DiffMethod

#### Enumeration Members

##### FORCE

```ts
FORCE: "FORCE";
```

###### Source

[config.ts:29](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L29)

##### NONE

```ts
NONE: "NONE";
```

###### Source

[config.ts:28](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L28)

## Variables

### ASSET\_HOMEPAGE\_LOCATION

```ts
const ASSET_HOMEPAGE_LOCATION: string;
```

#### Source

[config.ts:40](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L40)

***

### DEFAULT\_OPTIONS

```ts
const DEFAULT_OPTIONS: Readonly<Pick<ConfigOptions, "customDirective" | "groupByDirective" | "loaders"> & Required<Omit<ConfigOptions, "customDirective" | "groupByDirective" | "loaders">>>;
```

#### Source

[config.ts:47](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L47)

***

### DOCS\_URL

```ts
const DOCS_URL: "https://graphql-markdown.dev/docs";
```

#### Source

[config.ts:38](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L38)

***

### PACKAGE\_NAME

```ts
const PACKAGE_NAME: "@graphql-markdown/docusaurus";
```

#### Source

[config.ts:39](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L39)

## Functions

### buildConfig()

```ts
function buildConfig(
   configFileOpts, 
   cliOpts, 
id): Promise<Options>
```

#### Parameters

• **configFileOpts**: `Maybe`\<`ConfigOptions`\>

• **cliOpts**: `Maybe`\<`CliOptions`\>

• **id**: `Maybe`\<`string`\>= `"default"`

#### Returns

`Promise`\<`Options`\>

#### Source

[config.ts:330](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L330)

***

### getCustomDirectives()

```ts
function getCustomDirectives(customDirectiveOptions, skipDocDirective?): Maybe<CustomDirective>
```

#### Parameters

• **customDirectiveOptions**: `Maybe`\<`CustomDirective`\>

• **skipDocDirective?**: `Maybe`\<`DirectiveName`[]\>

#### Returns

`Maybe`\<`CustomDirective`\>

#### Source

[config.ts:170](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L170)

***

### getDiffMethod()

```ts
function getDiffMethod(diff, force): TypeDiffMethod
```

#### Parameters

• **diff**: `TypeDiffMethod`

• **force**: `boolean`= `false`

#### Returns

`TypeDiffMethod`

#### Source

[config.ts:203](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L203)

***

### getDocDirective()

```ts
function getDocDirective(name): DirectiveName
```

#### Parameters

• **name**: `Maybe`\<`DirectiveName`\>

#### Returns

`DirectiveName`

#### Source

[config.ts:86](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L86)

***

### getDocOptions()

```ts
function getDocOptions(cliOpts?, configOptions?): Required<ConfigDocOptions>
```

#### Parameters

• **cliOpts?**: `Maybe`\<`CliOptions` & `DeprecatedCliOptions`\>

• **configOptions?**: `Maybe`\<`ConfigDocOptions` & `DeprecatedConfigDocOptions`\>

#### Returns

`Required`\<`ConfigDocOptions`\>

#### Source

[config.ts:253](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L253)

***

### getOnlyDocDirectives()

```ts
function getOnlyDocDirectives(cliOpts, configFileOpts): DirectiveName[]
```

#### Parameters

• **cliOpts**: `Maybe`\<`CliOptions`\>

• **configFileOpts**: `Maybe`\<`Pick`\<`ConfigOptions`, `"onlyDocDirective"`\>\>

#### Returns

`DirectiveName`[]

#### Source

[config.ts:102](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L102)

***

### getPrintTypeOptions()

```ts
function getPrintTypeOptions(cliOpts, configOptions): Required<ConfigPrintTypeOptions>
```

#### Parameters

• **cliOpts**: `Maybe`\<`CliOptions`\>

• **configOptions**: `Maybe`\<`ConfigPrintTypeOptions`\>

#### Returns

`Required`\<`ConfigPrintTypeOptions`\>

#### Source

[config.ts:270](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L270)

***

### getSkipDocDirectives()

```ts
function getSkipDocDirectives(cliOpts, configFileOpts): DirectiveName[]
```

#### Parameters

• **cliOpts**: `Maybe`\<`CliOptions`\>

• **configFileOpts**: `Maybe`\<`Pick`\<`ConfigOptions`, `"printTypeOptions"` \| `"skipDocDirective"`\>\>

#### Returns

`DirectiveName`[]

#### Source

[config.ts:118](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L118)

***

### getVisibilityDirectives()

```ts
function getVisibilityDirectives(cliOpts, configFileOpts): object
```

#### Parameters

• **cliOpts**: `Maybe`\<`CliOptions`\>

• **configFileOpts**: `Maybe`\<`Pick`\<`ConfigOptions`, `"printTypeOptions"` \| `"skipDocDirective"` \| `"onlyDocDirective"`\>\>

#### Returns

`object`

##### onlyDocDirective

```ts
onlyDocDirective: DirectiveName[];
```

##### skipDocDirective

```ts
skipDocDirective: DirectiveName[];
```

#### Source

[config.ts:145](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L145)

***

### parseDeprecatedDocOptions()

```ts
function parseDeprecatedDocOptions(cliOpts, configOptions): Partial<object>
```

#### Parameters

• **cliOpts**: `Maybe`\<`DeprecatedCliOptions`\>

• **configOptions**: `Maybe`\<`DeprecatedConfigDocOptions`\>

#### Returns

`Partial`\<`object`\>

##### hide\_table\_of\_contents

```ts
hide_table_of_contents: boolean;
```

##### pagination\_next

```ts
pagination_next: null;
```

##### pagination\_prev

```ts
pagination_prev: null;
```

#### Source

[config.ts:212](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L212)

***

### parseGroupByOption()

```ts
function parseGroupByOption(groupOptions): Maybe<GroupByDirectiveOptions>
```

#### Parameters

• **groupOptions**: `unknown`

#### Returns

`Maybe`\<`GroupByDirectiveOptions`\>

#### Source

[config.ts:301](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/config.ts#L301)
