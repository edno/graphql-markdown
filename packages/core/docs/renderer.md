# renderer

## Classes

### Renderer

#### Constructors

##### new Renderer()

```ts
new Renderer(
   printer, 
   outputDir, 
   baseURL, 
   group, 
   prettify, 
   docOptions): Renderer
```

###### Parameters

• **printer**: *typeof* `IPrinter`

• **outputDir**: `string`

• **baseURL**: `string`

• **group**: `Maybe`\<`Partial`\<`Record`\<`SchemaEntity`, `Record`\<`string`, `Maybe`\<`string`\>\>\>\>\>

• **prettify**: `boolean`

• **docOptions**: `Maybe`\<`RendererDocOptions`\>

###### Returns

[`Renderer`](renderer.md#renderer)

###### Source

[renderer.ts:47](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L47)

#### Properties

##### baseURL

```ts
baseURL: string;
```

###### Source

[renderer.ts:41](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L41)

##### group

```ts
group: Maybe<Partial<Record<SchemaEntity, Record<string, Maybe<string>>>>>;
```

###### Source

[renderer.ts:39](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L39)

##### options

```ts
options: Maybe<RendererDocOptions>;
```

###### Source

[renderer.ts:43](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L43)

##### outputDir

```ts
outputDir: string;
```

###### Source

[renderer.ts:40](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L40)

##### prettify

```ts
prettify: boolean;
```

###### Source

[renderer.ts:42](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L42)

#### Methods

##### generateCategoryMetafile()

```ts
generateCategoryMetafile(
   category, 
   dirPath, 
   sidebarPosition?, 
   styleClass?, 
options?): Promise<void>
```

###### Parameters

• **category**: `string`

• **dirPath**: `string`

• **sidebarPosition?**: `number`

• **styleClass?**: `string`

• **options?**= `undefined`

• **options.collapsed?**: `boolean`

• **options.collapsible?**: `boolean`

###### Returns

`Promise`\<`void`\>

###### Source

[renderer.ts:63](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L63)

##### generateCategoryMetafileType()

```ts
generateCategoryMetafileType(
   type, 
   name, 
rootTypeName): Promise<string>
```

###### Parameters

• **type**: `unknown`

• **name**: `string`

• **rootTypeName**: `SchemaEntity`

###### Returns

`Promise`\<`string`\>

###### Source

[renderer.ts:98](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L98)

##### renderHomepage()

```ts
renderHomepage(homepageLocation): Promise<void>
```

###### Parameters

• **homepageLocation**: `string`

###### Returns

`Promise`\<`void`\>

###### Source

[renderer.ts:248](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L248)

##### renderRootTypes()

```ts
renderRootTypes(rootTypeName, type): Promise<Maybe<Maybe<Category>[]>>
```

###### Parameters

• **rootTypeName**: `SchemaEntity`

• **type**: `unknown`

###### Returns

`Promise`\<`Maybe`\<`Maybe`\<`Category`\>[]\>\>

###### Source

[renderer.ts:145](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L145)

##### renderSidebar()

```ts
renderSidebar(): Promise<string>
```

###### Returns

`Promise`\<`string`\>

###### Source

[renderer.ts:222](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L222)

##### renderTypeEntities()

```ts
renderTypeEntities(
   dirPath, 
   name, 
type): Promise<Maybe<Category>>
```

###### Parameters

• **dirPath**: `string`

• **name**: `string`

• **type**: `unknown`

###### Returns

`Promise`\<`Maybe`\<`Category`\>\>

###### Source

[renderer.ts:174](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/renderer.ts#L174)
