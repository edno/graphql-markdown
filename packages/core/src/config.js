const { join } = require("path");
const { tmpdir } = require("os");

const { parseGroupByOption } = require("./lib/group-info");

const PACKAGE_NAME = "@graphql-markdown/docusaurus";
const ASSETS_LOCATION = join(__dirname, "../assets/");

const DEFAULT_OPTIONS = {
  schema: "./schema.graphql",
  rootPath: "./docs",
  baseURL: "schema",
  linkRoot: "/",
  homepage: join(ASSETS_LOCATION, "generated.md"),
  diffMethod: undefined,
  tmpDir: join(tmpdir(), PACKAGE_NAME),
  loaders: {},
  pretty: false,
  printer: "@graphql-markdown/printer-legacy",
  docOptions: {
    pagination: true,
    toc: true,
    index: false,
  },
  printTypeOptions: {
    parentTypePrefix: true,
    relatedTypeSection: true,
    typeBadges: true,
  },
  skipDocDirective: undefined,
};

function buildConfig(configFileOpts, cliOpts) {
  let config = DEFAULT_OPTIONS;

  if (typeof configFileOpts !== "undefined" && configFileOpts != null) {
    config = { ...DEFAULT_OPTIONS, ...configFileOpts };
  }

  if (typeof cliOpts === "undefined" || cliOpts == null) {
    cliOpts = {};
  }

  const baseURL = cliOpts.base ?? config.baseURL;

  return {
    baseURL,
    schemaLocation: cliOpts.schema ?? config.schema,
    outputDir: join(cliOpts.root ?? config.rootPath, baseURL),
    linkRoot: cliOpts.link ?? config.linkRoot,
    homepageLocation: cliOpts.homepage ?? config.homepage,
    diffMethod: getDiffMethod(cliOpts.diff ?? config.diffMethod, cliOpts.force),
    tmpDir: cliOpts.tmp ?? config.tmpDir,
    loaders: config.loaders,
    groupByDirective:
      parseGroupByOption(cliOpts.groupByDirective) || config.groupByDirective,
    prettify: cliOpts.pretty ?? config.pretty,
    docOptions: getDocOptions(cliOpts, config.docOptions),
    printTypeOptions: gePrintTypeOptions(cliOpts, config.printTypeOptions),
    printer: config.printer,
    skipDocDirective: getSkipDocDirective(
      cliOpts.skip ?? config.skipDocDirective,
    ),
  };
}

function getDiffMethod(diff, force) {
  return force ? "FORCE" : diff;
}

function getDocOptions(cliOpts, configOptions) {
  return {
    pagination: !cliOpts.noPagination && configOptions.pagination,
    toc: !cliOpts.noToc && configOptions.toc,
    index: cliOpts.index || configOptions.index,
  };
}

function gePrintTypeOptions(cliOpts, configOptions) {
  return {
    parentTypePrefix: !cliOpts.noParentType && configOptions.parentTypePrefix,
    relatedTypeSection:
      !cliOpts.noRelatedType && configOptions.relatedTypeSection,
    typeBadges: !cliOpts.noTypeBadges && configOptions.typeBadges,
  };
}

function getSkipDocDirective(option) {
  const OPTION_REGEX = /^@(?<directive>\w+)$/;

  if (typeof option !== "string") {
    return undefined;
  }

  const parsedOption = OPTION_REGEX.exec(option);

  if (typeof parsedOption === "undefined" || parsedOption == null) {
    throw new Error(`Invalid "${option}"`);
  }

  return parsedOption.groups.directive;
}

module.exports = { buildConfig, DEFAULT_OPTIONS, ASSETS_LOCATION };
