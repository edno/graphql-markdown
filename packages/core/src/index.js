/* istanbul ignore file */

module.exports = {
  generateDocFromSchema: require("./lib/generator").generateDocFromSchema,
  groupInfo: require("./lib/group-info"),
  config: require("./config"),
};
