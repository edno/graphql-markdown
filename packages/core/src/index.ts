/* istanbul ignore file */

export { generateDocFromSchema } from "./generator";
export * from "./config";
