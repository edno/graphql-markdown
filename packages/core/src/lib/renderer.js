const path = require("path");

const {
  array: { convertArrayToObject },
  object: { hasProperty },
  string: { toSlug, startCase },
  url: { pathUrl },
} = require("@graphql-markdown/utils").scalars;

const {
  prettier: { prettifyJavascript, prettifyMarkdown },
  fs: { saveFile, ensureDir, copyFile, readFile, fileExists },
} = require("@graphql-markdown/utils").helpers;

const { ASSETS_LOCATION } = require("../config");
const { schemaSidebar } = require(`${ASSETS_LOCATION}/sidebar.json`);

const SIDEBAR = "sidebar-schema.js";
const HOMEPAGE_ID = "schema";
const CATEGORY_YAML = "_category_.yml";

module.exports = class Renderer {
  constructor(printer, outputDir, baseURL, group, prettify, docOptions) {
    this.group = group;
    this.outputDir = outputDir;
    this.baseURL = baseURL;
    this.printer = printer;
    this.prettify = prettify;
    this.options = docOptions;
  }

  async generateCategoryMetafile(category, dirPath) {
    const filePath = path.join(dirPath, CATEGORY_YAML);

    if (await fileExists(filePath)) {
      return;
    }

    await ensureDir(dirPath);

    const label = startCase(category);
    const link =
      typeof this.options === "undefined" || !this.options.index
        ? "null"
        : `\n  type: generated-index\n  title: '${label} overview'\n`;

    await saveFile(filePath, `label: ${label}\nlink: ${link}\n`);
  }

  async renderRootTypes(rootTypeName, type) {
    if (typeof type === "undefined" || type === null) {
      return undefined;
    }

    if (Array.isArray(type)) {
      type = convertArrayToObject(type);
    }

    return Promise.all(
      Object.keys(type).map(async (name) => {
        let dirPath = this.outputDir;

        if (hasProperty(this.group, name)) {
          dirPath = path.join(dirPath, toSlug(this.group[name]));
          await this.generateCategoryMetafile(this.group[name], dirPath);
        }

        dirPath = path.join(dirPath, toSlug(rootTypeName));
        await this.generateCategoryMetafile(rootTypeName, dirPath);

        return this.renderTypeEntities(dirPath, name, type[name]);
      }),
    );
  }

  async renderTypeEntities(dirPath, name, type) {
    if (typeof type === "undefined" || type === null) {
      return undefined;
    }

    const fileName = toSlug(name);
    const filePath = path.join(path.normalize(dirPath), `${fileName}.mdx`);

    const content = this.printer.printType(fileName, type, this.options);
    await saveFile(
      filePath,
      this.prettify ? prettifyMarkdown(content) : content,
    );

    const pagePath = path.relative(this.outputDir, filePath);
    const page = pagePath.match(
      /(?<category>[A-Za-z0-9-]+)[\\/]+(?<pageId>[A-Za-z0-9-]+).mdx?$/,
    );
    const slug = pathUrl.join(page.groups.category, page.groups.pageId);

    return { category: startCase(page.groups.category), slug: slug };
  }

  async renderSidebar() {
    const sidebar = {
      schemaSidebar: schemaSidebar.map((entry) => {
        switch (entry.type) {
          case "doc":
            entry.id = pathUrl.join(this.baseURL, HOMEPAGE_ID);
            break;
          case "autogenerated":
            entry.dirName = this.baseURL;
            break;
          default: //do nothing
        }
        return entry;
      }),
    };

    const jsonSidebar = `
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */

module.exports = ${JSON.stringify(sidebar, null, 2)};
`;

    const filePath = path.join(this.outputDir, SIDEBAR);
    await saveFile(
      filePath,
      this.prettify ? prettifyJavascript(jsonSidebar) : jsonSidebar,
    );

    return path.relative("./", filePath);
  }

  async renderHomepage(homepageLocation) {
    const homePage = path.basename(homepageLocation);
    const destLocation = path.join(this.outputDir, homePage);
    const slug = pathUrl.resolve("/", this.baseURL);

    await copyFile(homepageLocation, destLocation);

    const template = await readFile(destLocation);

    const data = template
      .toString()
      .replace(/##baseURL##/gm, slug)
      .replace(/##generated-date-time##/gm, new Date().toLocaleString());
    await saveFile(destLocation, data);
  }
};
