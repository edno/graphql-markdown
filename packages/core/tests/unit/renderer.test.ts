jest.mock("node:fs/promises");

import path from "node:path";
jest.mock("node:path", () => jest.requireActual("node:path"));

import { GraphQLScalarType, Kind } from "graphql";

import type {
  IPrinter,
  MDXString,
  SchemaEntity,
  TypeDeprecatedOption,
} from "@graphql-markdown/types";

jest.mock("@graphql-markdown/printer-legacy");
import { Printer } from "@graphql-markdown/printer-legacy";

jest.mock("@graphql-markdown/utils", () => {
  return {
    __esModule: true,
    ...jest.requireActual("@graphql-markdown/utils"),
    isDeprecated: jest.fn(),
    ensureDir: jest.fn(),
    fileExists: jest.fn(),
    saveFile: jest.fn(),
    copyFile: jest.fn(),
  };
});
import * as Utils from "@graphql-markdown/utils";

import { Renderer } from "../../src/renderer";
import { DEFAULT_OPTIONS } from "../../src/config";

const DEFAULT_RENDERER_OPTIONS = {
  ...DEFAULT_OPTIONS.docOptions,
  deprecated: DEFAULT_OPTIONS.printTypeOptions
    .deprecated as TypeDeprecatedOption,
};

describe("renderer", () => {
  describe("class Renderer", () => {
    let rendererInstance: Renderer;
    const baseURL: string = "graphql";

    beforeEach(() => {
      rendererInstance = new Renderer(
        Printer as unknown as typeof IPrinter,
        "/output",
        baseURL,
        undefined,
        DEFAULT_OPTIONS.pretty,
        {
          ...DEFAULT_OPTIONS.docOptions,
          deprecated: DEFAULT_OPTIONS.printTypeOptions
            .deprecated as TypeDeprecatedOption,
        },
      );
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    describe("renderTypeEntities()", () => {
      test("creates entity page into output folder", async () => {
        expect.assertions(2);

        jest
          .spyOn(Printer, "printType")
          .mockReturnValue("Lorem ipsum" as MDXString);
        const spy = jest.spyOn(Utils, "saveFile");

        const output = "/output/foobar";
        const meta = await rendererInstance.renderTypeEntities(
          output,
          "FooBar",
          "FooBar",
        );

        expect(meta).toEqual({ category: "Foobar", slug: "foobar/foo-bar" });
        expect(spy).toHaveBeenCalledWith(
          `${output}/foo-bar.mdx`,
          "Lorem ipsum",
          undefined,
        );
      });

      test("do nothing if type is not defined", async () => {
        expect.assertions(1);

        jest.spyOn(Printer, "printType").mockReturnValue(undefined);

        const meta = await rendererInstance.renderTypeEntities(
          "test",
          "FooBar",
          null,
        );

        expect(meta).toBeUndefined();
      });

      test("return undefined and logs warning if an error is thrown", async () => {
        expect.assertions(2);

        jest.spyOn(Printer, "printType").mockImplementationOnce(() => {
          throw new Error();
        });

        const spy = jest
          .spyOn(global.console, "warn")
          .mockImplementationOnce(() => {});

        const meta = await rendererInstance.renderTypeEntities(
          "test",
          "FooBar",
          "TestType",
        );

        expect(meta).toBeUndefined();
        expect(spy).toHaveBeenCalledWith(
          `An error occurred while processing "TestType"`,
        );
      });

      test("return undefined and logs warning if file path is invalid", async () => {
        expect.assertions(2);

        jest
          .spyOn(Printer, "printType")
          .mockReturnValue("Lorem ipsum" as MDXString);
        jest
          .spyOn(Utils.pathUrl, "relative")
          .mockImplementationOnce(() => "not-valid.md");
        const spy = jest
          .spyOn(global.console, "warn")
          .mockImplementationOnce(() => {});

        const meta = await rendererInstance.renderTypeEntities(
          "test",
          "FooBar",
          "TestType",
        );

        expect(meta).toBeUndefined();
        expect(spy).toHaveBeenCalledWith(
          `An error occurred while processing file test/foo-bar.mdx for type "TestType"`,
        );
      });
    });

    describe("renderSidebar()", () => {
      test("creates Docusaurus compatible sidebar.js into output folder", async () => {
        expect.assertions(2);

        const spy = jest.spyOn(Utils, "saveFile");

        const filePath = await rendererInstance.renderSidebar();

        expect(spy.mock.calls).toMatchSnapshot();
        expect(filePath.endsWith("output/sidebar-schema.js")).toBeTruthy();
      });
    });

    describe("renderHomepage()", () => {
      test("copies default homepage into output folder", async () => {
        expect.assertions(1);

        jest.spyOn(Utils, "readFile").mockResolvedValueOnce("Test Homepage");
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.renderHomepage("/assets/generated.md");

        expect(spy).toHaveBeenCalledWith(
          "/output/generated.md",
          "Test Homepage",
        );
      });
    });

    describe("renderRootTypes()", () => {
      test("render root type", async () => {
        expect.assertions(2);

        jest
          .spyOn(Printer, "printType")
          .mockImplementation(() => "content" as MDXString);
        jest.spyOn(Utils, "fileExists").mockResolvedValue(true);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.renderRootTypes("objects", {
          foo: new GraphQLScalarType({
            name: "foo",
            astNode: {
              kind: Kind.SCALAR_TYPE_DEFINITION,
              name: { kind: Kind.NAME, value: "foo" },
            },
          }),
          bar: new GraphQLScalarType({
            name: "bar",
            astNode: {
              kind: Kind.SCALAR_TYPE_DEFINITION,
              name: { kind: Kind.NAME, value: "foo" },
            },
          }),
        });

        expect(spy).toHaveBeenNthCalledWith(
          1,
          "/output/objects/foo.mdx",
          "content",
          undefined,
        );
        expect(spy).toHaveBeenNthCalledWith(
          2,
          "/output/objects/bar.mdx",
          "content",
          undefined,
        );
      });
    });

    describe("generateCategoryMetafile()", () => {
      test("generate _category_.yml file", async () => {
        expect.assertions(1);

        const category = "foobar";
        const outputPath = "/output/docs";
        const filePath = path.join(outputPath, "_category_.yml");

        jest.spyOn(Utils, "fileExists").mockResolvedValue(false);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.generateCategoryMetafile(category, outputPath);

        expect(spy).toHaveBeenCalledWith(
          filePath,
          `label: Foobar\nposition: 1\nlink: null\n`,
        );
      });

      test("generate _category_.yml file with generated index", async () => {
        expect.assertions(1);

        const category = "foobar";
        const outputPath = "/output/docs";
        const filePath = path.join(outputPath, "_category_.yml");

        rendererInstance.options = {
          ...DEFAULT_RENDERER_OPTIONS,
          index: true,
        };

        jest.spyOn(Utils, "fileExists").mockResolvedValue(false);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.generateCategoryMetafile(category, outputPath);

        expect(spy).toHaveBeenCalledWith(
          filePath,
          `label: Foobar\nposition: 1\nlink: \n  type: generated-index\n  title: 'Foobar overview'\n`,
        );
      });

      test("do not generate _category_.yml file if it exists", async () => {
        expect.assertions(1);

        const category = "foobar";
        const outputPath = "/output/docs";

        jest.spyOn(Utils, "fileExists").mockResolvedValue(true);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.generateCategoryMetafile(category, outputPath);

        expect(spy).not.toHaveBeenCalled();
      });

      test("generate _category_.yml file with sidebar position", async () => {
        expect.assertions(1);

        const category = "foobar";
        const outputPath = "/output/docs";
        const filePath = path.join(outputPath, "_category_.yml");

        jest.spyOn(Utils, "fileExists").mockResolvedValue(false);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.generateCategoryMetafile(
          category,
          outputPath,
          42,
        );

        expect(spy).toHaveBeenCalledWith(
          filePath,
          `label: Foobar\nposition: 42\nlink: null\n`,
        );
      });

      test("generate _category_.yml file with classname", async () => {
        expect.assertions(1);

        const category = "foobar";
        const outputPath = "/output/docs";
        const filePath = path.join(outputPath, "_category_.yml");
        const styleClass = "foo-baz";

        jest.spyOn(Utils, "fileExists").mockResolvedValue(false);
        const spy = jest.spyOn(Utils, "saveFile");

        await rendererInstance.generateCategoryMetafile(
          category,
          outputPath,
          42,
          styleClass,
        );

        expect(spy).toHaveBeenCalledWith(
          filePath,
          `label: Foobar\nposition: 42\nclassName: ${styleClass}\nlink: null\n`,
        );
      });
    });

    describe("generateCategoryMetafileType()", () => {
      test("generate type _category_.yml file", async () => {
        expect.assertions(2);

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          {},
          "Foo",
          "objects",
        );

        expect(spy).toHaveBeenCalledWith("objects", "/output/objects");
        expect(dirPath).toBe("/output/objects");
      });

      test("generate group _category_.yml file", async () => {
        expect.assertions(3);

        const [type, name, root, group] = [{}, "Foo", "objects", "lorem"];

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");
        jest.replaceProperty(rendererInstance, "group", {
          [root]: { [name]: group },
        });

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          type,
          name,
          root as SchemaEntity,
        );

        expect(spy).toHaveBeenCalledWith("lorem", "/output/lorem");
        expect(spy).toHaveBeenCalledWith("objects", "/output/lorem/objects");
        expect(dirPath).toBe(`/output/${group}/${root.toLowerCase()}`);
      });

      test("generate deprecated _category_.yml file if type is deprecated", async () => {
        expect.assertions(3);

        const [type, name, root] = [{}, "Foo", "Baz"];

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");
        jest.replaceProperty(rendererInstance, "options", {
          deprecated: "group",
        });
        jest.spyOn(Utils, "isDeprecated").mockReturnValue(true);

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          type,
          name,
          root as SchemaEntity,
        );

        expect(spy).toHaveBeenCalledWith(
          "deprecated",
          "/output/deprecated",
          999,
          "deprecated",
        );
        expect(spy).toHaveBeenCalledWith("Baz", "/output/deprecated/baz");
        expect(dirPath).toBe(`/output/deprecated/${root.toLowerCase()}`);
      });

      test("generate no deprecated _category_.yml file if type is not deprecated", async () => {
        expect.assertions(2);

        const [type, name, root] = [{}, "Foo", "Baz"];

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");
        jest.replaceProperty(rendererInstance, "options", {
          deprecated: "group",
        });
        jest.spyOn(Utils, "isDeprecated").mockReturnValue(false);

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          type,
          name,
          root as SchemaEntity,
        );

        expect(spy).toHaveBeenCalledWith("Baz", "/output/baz");
        expect(dirPath).toBe(`/output/${root.toLowerCase()}`);
      });

      test("generate no deprecated _category_.yml file if deprecated is not group", async () => {
        expect.assertions(2);

        const [type, name, root] = [{}, "Foo", "Baz"];

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");
        jest.replaceProperty(rendererInstance, "options", {
          deprecated: "default",
        });
        jest.spyOn(Utils, "isDeprecated").mockReturnValue(false);

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          type,
          name,
          root as SchemaEntity,
        );

        expect(spy).toHaveBeenCalledWith("Baz", "/output/baz");
        expect(dirPath).toBe(`/output/${root.toLowerCase()}`);
      });

      test("generate deprecated and group _category_.yml files", async () => {
        expect.assertions(2);

        const [type, name, root, group] = [{}, "Foo", "Baz", "lorem"];

        const spy = jest.spyOn(rendererInstance, "generateCategoryMetafile");
        jest.replaceProperty(rendererInstance, "options", {
          deprecated: "group",
        });
        jest.replaceProperty(rendererInstance, "group", {
          [root]: { [name]: group },
        });
        jest.spyOn(Utils, "isDeprecated").mockReturnValue(true);

        const dirPath = await rendererInstance.generateCategoryMetafileType(
          type,
          name,
          root as SchemaEntity,
        );

        expect(spy).toHaveBeenCalledTimes(3);
        expect(dirPath).toBe(
          `/output/deprecated/${group}/${root.toLowerCase()}`,
        );
      });
    });
  });
});
