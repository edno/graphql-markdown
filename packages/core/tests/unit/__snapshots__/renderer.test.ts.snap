// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer class Renderer renderSidebar() creates Docusaurus compatible sidebar.js into output folder 1`] = `
[
  [
    "/output/sidebar-schema.js",
    "
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */

module.exports = {
  "schemaSidebar": [
    {
      "type": "autogenerated",
      "dirName": "graphql"
    }
  ]
};
",
    undefined,
  ],
]
`;
